external-crates/move/crates/move-compiler/src/typing/syntax_methods.rs

```diff
@@ -95,64 +93,91 @@ fn validate_index_syntax_methods(
-        let id = N::TParamID::next();
-        let user_specified_name = sp(index_tyarg.loc, format!("_{}", ndx).into());
-        let tparam = N::TParam {
-            id,
-            user_specified_name,
-            abilities: AbilitySet::all(index_ann_loc),
-        };
-        if let Ok((subst_, _)) = core::join(
-            subst.clone(),
-            index_tyarg,
-            &sp(index_tyarg.loc, N::Type_::Param(tparam)),
-        ) {
-            subst = subst_;
-        } else {
-            context
-                .env
-                .add_diag(ice!((index.loc, "Failed in validating syntax methods")));
-            valid = false;
+
+        for ability in mut_tparam.abilities.iter() {
+            if !index_tparam.abilities.has_ability(&ability) {
+                let index_mut_msg = format!(
+                    "This mutable index function type parameter has the '{}' ability",
+                    ability
+                );
+                let index_msg = "This index function type parameter does not";
+                let mut diag = diag!(
+                    TypeSafety::IncompatibleSyntaxMethods,
+                    (mut_tparam.user_specified_name.loc, index_mut_msg),
+                    (ability.loc, "Ability defined here"),
+                    (index_tparam.user_specified_name.loc, index_msg),
+                );
+                diag.add_note(
+                    "Index operations on use the same abilities for their type parameters",
+                );
+                context.env.add_diag(diag);
+                valid = false;
+            }
         }
     }
 
+    // Now we simply want to make the types the same w/r/t type parameters. To do this, we
+    // instantiate the type parameters of the index one and ground them to the type parameters of
+    // the mutable one. Finally, we walk both types and make sure they agree.
+
+    // make_function_type updates the subst on the context, and we also don't want to leave
+    // lingering constraints, so we preserve the current versions here to reinstate at the end.
+    let prev_constraints = std::mem::take(&mut context.constraints);
+    let prev_subst = std::mem::replace(&mut context.subst, core::Subst::empty());
+
+    let mut_tparam_types = mut_finfo
+        .signature
+        .type_parameters
+        .iter()
+        .map(|tp| sp(tp.user_specified_name.loc, N::Type_::Param(tp.clone())))
+        .collect::<Vec<_>>();
+
+    let index_ty = core::make_function_type(
+        context,
+        index_ann_loc,
+        index_module,
+        index_fn,
+        Some(mut_tparam_types),
+    );
+
     let mut param_tys = index_ty
         .params
         .iter()
-        .zip(index_mut_ty.params.iter())
-        .map(|((_, t1), (_, t2))| (t1, t2))
+        .map(|(_, t1)| t1)
+        .zip(mut_finfo.signature.parameters.iter().map(|(_, _, ty)| ty))
         .enumerate();
 
+    let mut subst = std::mem::replace(&mut context.subst, core::Subst::empty());
+
```

Comment:
What's this one for?

Reply:
I need a substitution to hand into the impls for invariant and subtyping, and this substitution has the function instantiation on it.